viewing_party/party.py

```diff
@@ -1,23 +1,201 @@
+    
+    #dictionary with genres as keys and amount of movies of that genre watched as values by counting times genre is in genres_watched list
+    for genre in genres_list:
+        watched_by_genre[genre] = genres_list.count(genre)
+    
+    top_times_watched = 0
+    top_genre = None
+    
+    # looping through dictionary to find most popular genre
+    for genre in watched_by_genre:
+        times_watched = watched_by_genre[genre]
+        if times_watched > top_times_watched:
+            top_times_watched = times_watched
+            top_genre = genre
+    
+    return top_genre
+    
+
+
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    
+    friends_watched_movies = set()
+    user_watched_movies = set()
+
+    # creates a set of friends' watched movies
+    for friend in user_data["friends"]:
+        for watched_list in friend["watched"]:
+            friends_watched_movies.add(watched_list['title'])
+    
+    # creates a set of user's watched movies
+    for movie in user_data["watched"]:
+        user_watched_movies.add(movie["title"])
+
+    unique_movie_titles = (user_watched_movies - friends_watched_movies)
+
+    unique_movies_data = []
+
+    #looks for unique movies in user_data and adds movie data dictionary to unique_movies_data. this could be turned into a helper function
+    for movie in unique_movie_titles:
+        for i in range(len(user_data["watched"])):
+            if movie == user_data["watched"][i]["title"]:
+                unique_movies_data.append(user_data["watched"][i])
+
+    return unique_movies_data
+
+
+def get_friends_unique_watched(user_data):
+    friends_watched_movies = set()
+    user_watched_movies = set()
+    
+    for friend in user_data["friends"]:
+        for watched_list in friend["watched"]:
+            friends_watched_movies.add(watched_list['title'])
+
+    for movie in user_data["watched"]:
+        user_watched_movies.add(movie["title"])
+
+    unique_movie_titles = (friends_watched_movies - user_watched_movies)
+
+    unique_movies_data = []
+
+    for movie in unique_movie_titles:
+        for friend in (user_data["friends"]):
+            for i in range(len(friend["watched"])):
+                if movie == friend["watched"][i]["title"] and friend["watched"][i] not in unique_movies_data:
+                    unique_movies_data.append(friend["watched"][i])
+
+    return unique_movies_data
+
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    available_recs = []
+    user_subscriptions = user_data.get("subscriptions", [])
+
+    for friends_list in user_data["friends"]:
+        for movie in friends_list["watched"]:
+            if movie["host"] in user_subscriptions and movie not in user_data["watched"]:

```

Comment:
We could avoid checking whether the movie is in the `user_data` (`in` on a list is slower than for a set) if we start our iteration from `get_friends_unique_watched`, since that already eliminates movies that the user has watched.